server/src/main/java/org/apache/druid/segment/metadata/AbstractSegmentMetadataCache.java

```diff
@@ -732,18 +718,35 @@ public Set<SegmentId> refreshSegmentsForDataSource(final String dataSource, fina
 
     log.debug("Refreshing metadata for datasource[%s].", dataSource);
 
+    final Set<SegmentId> retVal = new HashSet<>();
+
+    ConcurrentSkipListMap<SegmentId, AvailableSegmentMetadata> datasourceSegments = segmentMetadataInfo.get(dataSource);
+    // this datasource no longer exists, skip refresh
+    if (datasourceSegments == null) {
+      return retVal;
+    }
+
+    // Skip refreshing tombstone segments. These segments lack data or column information.
+    // Additionally, segment metadata queries, which are not yet implemented for tombstone segments
+    // (see: https://github.com/apache/druid/pull/12137) do not provide metadata for tombstones,
+    // leading to indefinite refresh attempts for these segments.
+    Set<SegmentId> segmentsWithoutTombstone =
```

Comment:
Why do we need to materialize this. 
Why can't we return a iterable which just skips the tombstone segments ?
We can always increment counters there no ?

Reply:
I have refactored to ensure that the stream is not materialized. 
For `count` I have added a terminal operation but that wouldn't materialize the entire stream.

Reply:
I just changed the approach yesterday. Instead of filtering out the tombstone segments in the end before refresh, I am ensuring there are never marked for refresh. 

A segment is marked for refresh in following scenarios:
* Segment is added. 
* Datasource signature is built and schema for segment is missing. 
* Metadata for the segment is fetched and schema for the segment is missing. 

I have ensured that a tombstone segment never gets marked for refresh itself.